velox/functions/prestosql/aggregates/ReduceAgg.cpp

```diff
@@ -80,40 +49,92 @@ using ReduceAggAccumulator = functions::aggregate::SingleValueAccumulator;
 //
 //   Combine 100 states into 50:
 //
-//   [s1, s2,...s50] = g([s1, s3,...s99], [s2, s4,...s100])
+//   [s1, s2,...s50] = g([s1, s2,...s50], [s51, s52,...s100])
 //
 //   Combine these 50 states into 25 states:
 //
-//   [s1, s2,...s25] = g([s1, s3,...s49], [s2, s4,...s50])
+//   [s1, s2,...s25] = g([s1, s2,...s25], [s26, s27,...s50])
+//
+//   Combine these 25 states into 13 (12 + 1) states:
+//
+//   [s1, s2,...s12] = g([s1, s2,...s12], [s13, s14,...s24])
+//   s13 = s25
 //
 //   Continue in this manner until all states are combined. This requires only
 //   log2(100), ~7, expression evaluations.
 //
-// Note that the more efficient algorithm for global aggregation doesn't support
-// applying reduce_agg to sorted inputs.
+//   Note: When adding a batch of rows to a non-empty accumulator, the
+//   accumulated state is appended to the list of initial states. In the
+//   example above, we would get 101 states to combine.
+//
+//   Group by applies this technique to each group.
+//
+//   Consider dataset {1, 10, 2, 20, 3} where values 1, 2, 3 belong to group 1
+//   and values 10, 20 belong to group 2.
+//
+//   The original implementation went like so:
+//
+//   s0 - initialValue, s1 - state for group 1, s2 - state for group 2.
+//
+//   s1 = s0
+//   s2 = s0
+//   s1 = f(s1, 1)
+//   s2 = f(s2, 10)
+//   s1 = f(s1, 2)
+//   s2 = f(s2, 20)
+//   s1 = f(s1, 3)
+//
+//   The inputFunction lambda expression was evaluated 5 times, once per row.
+//
+//   This implementation converts all inputs into states using inputFunction
+//   and initialValue, then combines multiple pairs of states similar to global
+//   aggregation.
 //
-// Also, note that the more efficient algorithm for global aggregation can be
-// used in a group-by as well. Each evaluation of the lambda could be reducing
-// the cardinality of each group by a factor of 2. This would be require
-// log2(max-group-cardinality) evaluations.
+//   First, put rows that belong to same groups together.
 //
-// A common use case for reduce_agg is to compute a product of input values.
+//   [1, 2, 3, 10, 30]
+//
+//   Convert values into states:
+//
+//   [s1, s2, s3, s4, s5] -> f([s0, s0, s0, s0, s0], [1, 2, 3, 10, 30])
+//
+//   Combine states within each group:
+//
+//   [s1, s2] = g([s1, s4], [s2, s5])
+//   s3 = s3 // extra state for group 1
+//
+//   We have 2 states for group 1 (s1, s3) and 1 state for group 2 (s2).
+//   Group 2 is done. We proceed to combine states for group 1.
+//
+//   s1 = g([s1], [s3])
+//
+//   In this example, inputFunction lambda expression is evaluated once and
+//   combineFunction is evaluated twice (compared to evaluating inputFunction 5
+//   times above).
+//
+//   Note: When adding a batch of rows to non-empty accumulators, the
+//   accumulated states are added to the list of initial states above.
+//
+// Note: This more efficient algorithm doesn't support applying reduce_agg
+// to sorted inputs.
```

Comment:
I am just curious why we don't support sorted inputs? Thanks!

Reply:
To preserve sort order of the inputs, we have to evaluate inputFunction on one value at a time. Unless, both inputFunction and combineFunction have this property:

g(g(g(a, b), c), d) == g(g(a, b), g(c, d))

Reply:
I see. Thanks for explanation!